File: js/src/internal/adf/markdown.ts
import type { ADFDocument, ADFNode, ADFMark } from '../types.js';

const PANEL_TYPES = new Set(['info', 'warning', 'note', 'success', 'error']);

export function markdownToADF(md: string): ADFDocument {
  const lines = md.split('\n');
  const content: ADFNode[] = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    if (line.startsWith('```')) {
      const lang = line.slice(3).trim();
      const codeLines: string[] = [];
      i++;
      while (i < lines.length && !lines[i].startsWith('```')) {
        codeLines.push(lines[i]);
        i++;
      }
      i++;
      const node: ADFNode = {
        type: 'codeBlock',
        content: [{ type: 'text', text: codeLines.join('\n') }],
      };
      if (lang) node.attrs = { language: lang };
      content.push(node);
      continue;
    }

    const headingMatch = line.match(/^(#{1,6})\s+(.*)/);
    if (headingMatch) {
      const level = headingMatch[1].length;
      content.push({
        type: 'heading',
        attrs: { level },
        content: parseInline(headingMatch[2]),
      });
      i++;
      continue;
    }

    const panelMatch = line.match(/^>\s*\[!(\w+)\]\s*(.*)/);
    if (panelMatch) {
      const rawType = panelMatch[1].toLowerCase();
      const panelType = PANEL_TYPES.has(rawType) ? rawType : 'info';
      const panelLines: string[] = [];
      const firstLine = panelMatch[2];
      if (firstLine.trim() !== '') panelLines.push(firstLine);
      i++;
      while (i < lines.length && lines[i].startsWith('>')) {
        panelLines.push(lines[i].replace(/^>\s?/, ''));
        i++;
      }
      content.push({
        type: 'panel',
        attrs: { panelType },
        content: [{ type: 'paragraph', content: parseInline(panelLines.join(' ')) }],
      });
      continue;
    }

    if (line.includes('|') && line.trim() !== '') {
      const next = lines[i + 1] ?? '';
      // A real Markdown table separator row must contain a `|` AND consist only of
      // pipes/dashes/colons/whitespace. Without the `|` requirement, a bullet list
      // line like `- foo` after a `|`-containing line gets misparsed as a table.
      const hasSeparator = next.includes('|') && /^[\s|:-]+$/.test(next);
      const nextIsTableRow = next.includes('|') && next.trim() !== '';
      if (hasSeparator || nextIsTableRow) {
        const tableRows: ADFNode[] = [];
        const headerCells = splitTableRow(line);
        tableRows.push({
          type: 'tableRow',
          content: headerCells.map((cell) => ({
            type: 'tableHeader',
            content: [{ type: 'paragraph', content: parseInline(cell.trim()) }],
          })),
        });
        i += hasSeparator ? 2 : 1;
        while (i < lines.length && lines[i].includes('|') && lines[i].trim() !== '') {
          const cells = splitTableRow(lines[i]);
          tableRows.push({
            type: 'tableRow',
            content: cells.map((cell) => ({
              type: 'tableCell',
              content: [{ type: 'paragraph', content: parseInline(cell.trim()) }],
            })),
          });
          i++;
        }
        content.push({ type: 'table', content: tableRows });
        continue;
      }
    }

    if (/^(-{3,}|\*{3,}|_{3,})$/.test(line.trim())) {
      content.push({ type: 'rule' });
      i++;
      continue;
    }

    if (/^\s*[-*]\s+/.test(line)) {
      const items: ADFNode[] = [];
      while (i < lines.length && /^\s*[-*]\s+/.test(lines[i])) {
        const text = lines[i].replace(/^\s*[-*]\s+/, '');
        items.push({
          type: 'listItem',
          content: [{ type: 'paragraph', content: parseInline(text) }],
        });
        i++;
      }
      content.push({ type: 'bulletList', content: items });
      continue;
    }

    if (/^\s*\d+\.\s+/.test(line)) {
      const items: ADFNode[] = [];
      while (i < lines.length && /^\s*\d+\.\s+/.test(lines[i])) {
        const text = lines[i].replace(/^\s*\d+\.\s+/, '');
        items.push({
          type: 'listItem',
          content: [{ type: 'paragraph', content: parseInline(text) }],
        });
        i++;
      }
      content.push({ type: 'orderedList', content: items });
      continue;
    }

    if (line.trim() === '') {
      i++;
      continue;
    }

    const paraLines: string[] = [line];
    i++;
    while (
      i < lines.length &&
      lines[i].trim() !== '' &&
      !lines[i].startsWith('```') &&
      !/^#{1,6}\s/.test(lines[i]) &&
      !/^\s*[-*]\s+/.test(lines[i]) &&
      !/^\s*\d+\.\s+/.test(lines[i]) &&
      !/^>/.test(lines[i]) &&
      !/^(-{3,}|\*{3,}|_{3,})$/.test(lines[i].trim()) &&
      !lines[i].includes('|')
    ) {
      paraLines.push(lines[i]);
      i++;
    }
    const paraContent: ADFNode[] = [];
    paraLines.forEach((pl, idx) => {
      if (idx > 0) paraContent.push({ type: 'hardBreak' });
      paraContent.push(...parseInline(pl));
    });
    content.push({ type: 'paragraph', content: paraContent });
  }

  if (content.length === 0) {
    content.push({ type: 'paragraph', content: [{ type: 'text', text: '' }] });
  }

  return { type: 'doc', version: 1, content };
}

function splitTableRow(row: string): string[] {
  let s = row.trim();
  if (s.startsWith('|')) s = s.slice(1);
  if (s.endsWith('|')) s = s.slice(0, -1);
  return s.split('|');
}

function isWordChar(ch: string | undefined): boolean {
  return ch !== undefined && /[A-Za-z0-9]/.test(ch);
}

function parseInline(text: string): ADFNode[] {
  const nodes: ADFNode[] = [];
  let i = 0;
  const len = text.length;

  while (i < len) {
    if (text[i] === '`') {
      const end = text.indexOf('`', i + 1);
      if (end !== -1) {
        nodes.push({ type: 'text', text: text.slice(i + 1, end), marks: [{ type: 'code' }] });
        i = end + 1;
        continue;
      }
    }

    if (text.slice(i, i + 2) === '**') {
      const end = text.indexOf('**', i + 2);
      if (end !== -1 && end > i + 2) {
        nodes.push({ type: 'text', text: text.slice(i + 2, end), marks: [{ type: 'strong' }] });
        i = end + 2;
        continue;
      }
    }

    if (text.slice(i, i + 2) === '__') {
      const before = text[i - 1];
      if (!isWordChar(before)) {
        const end = findDoubleUnderscoreEnd(text, i + 2);
        if (end !== -1 && end > i + 2) {
          nodes.push({ type: 'text', text: text.slice(i + 2, end), marks: [{ type: 'strong' }] });
          i = end + 2;
          continue;
        }
      }
    }

    if (text[i] === '*' && text[i + 1] !== '*' && text[i - 1] !== '*') {
      const end = findItalicEndAsterisk(text, i + 1);
      if (end !== -1 && end > i + 1) {
        nodes.push({ type: 'text', text: text.slice(i + 1, end), marks: [{ type: 'em' }] });
        i = end + 1;
        continue;
      }
    }

    if (text[i] === '_' && text[i + 1] !== '_' && !isWordChar(text[i - 1])) {
      const end = findItalicEndUnderscore(text, i + 1);
      if (end !== -1 && end > i + 1) {
        nodes.push({ type: 'text', text: text.slice(i + 1, end), marks: [{ type: 'em' }] });
        i = end + 1;
        continue;
      }
    }

    if (text.slice(i, i + 9) === '@mention(') {
      const end = text.indexOf(')', i + 9);
      if (end !== -1) {
        const inner = text.slice(i + 9, end);
        const commaIdx = inner.indexOf(',');
        if (commaIdx !== -1) {
          const id = inner.slice(0, commaIdx).trim();
          const name = inner.slice(commaIdx + 1).trim();
          nodes.push({
            type: 'mention',
            attrs: { id, text: '@' + name },
          });
          i = end + 1;
          continue;
        }
      }
    }

    if (text[i] === '[') {
      const closeBracket = text.indexOf(']', i + 1);
      if (closeBracket !== -1 && text[closeBracket + 1] === '(') {
        const closeParen = text.indexOf(')', closeBracket + 2);
        if (closeParen !== -1) {
          const linkText = text.slice(i + 1, closeBracket);
          const href = text.slice(closeBracket + 2, closeParen);
          const marks: ADFMark[] = [{ type: 'link', attrs: { href } }];
          nodes.push({ type: 'text', text: linkText, marks });
          i = closeParen + 1;
          continue;
        }
      }
    }

    let end = i + 1;
    while (end < len) {
      const ch = text[end];
      if (ch === '`' || ch === '[' || ch === '@') break;
      if (text.slice(end, end + 2) === '**') break;
      if (text.slice(end, end + 2) === '__' && !isWordChar(text[end - 1])) break;
      if (ch === '*' && text[end + 1] !== '*' && text[end - 1] !== '*') break;
      if (ch === '_' && text[end + 1] !== '_' && !isWordChar(text[end - 1])) break;
      end++;
    }
    nodes.push({ type: 'text', text: text.slice(i, end) });
    i = end;
  }

  if (nodes.length === 0) {
    nodes.push({ type: 'text', text: '' });
  }

  return nodes;
}

function findItalicEndAsterisk(text: string, start: number): number {
  for (let j = start; j < text.length; j++) {
    if (text[j] === '*' && text[j + 1] !== '*' && text[j - 1] !== '*') {
      return j;
    }
  }
  return -1;
}

function findItalicEndUnderscore(text: string, start: number): number {
  for (let j = start; j < text.length; j++) {
    if (text[j] === '_' && text[j + 1] !== '_' && !isWordChar(text[j + 1])) {
      return j;
    }
  }
  return -1;
}

function findDoubleUnderscoreEnd(text: string, start: number): number {
  for (let j = start; j < text.length - 1; j++) {
    if (text[j] === '_' && text[j + 1] === '_' && !isWordChar(text[j + 2])) {
      return j;
    }
  }
  return -1;
}

File: internal/adf/markdown.go
package adf

import (
	"strings"
	"unicode/utf8"
)

func markdownToADF(markdown string) map[string]interface{} {
	lines := strings.Split(markdown, "\n")
	content := []interface{}{}
	var codeBlock []string
	codeLang := ""
	inCode := false

	flushParagraph := func(text string) {
		if text == "" {
			return
		}
		inlineContent := parseInlineMarks(text)
		if len(inlineContent) > 0 {
			content = append(content, map[string]interface{}{
				"type":    "paragraph",
				"content": inlineContent,
			})
		}
	}

	for i := 0; i < len(lines); i++ {
		line := lines[i]

		if strings.HasPrefix(line, "```") {
			if inCode {
				content = append(content, map[string]interface{}{
					"type": "codeBlock",
					"attrs": map[string]interface{}{
						"language": codeLang,
					},
					"content": []interface{}{
						map[string]interface{}{
							"type": "text",
							"text": strings.Join(codeBlock, "\n"),
						},
					},
				})
				codeBlock = nil
				inCode = false
			} else {
				inCode = true
				codeLang = strings.TrimPrefix(line, "```")
				codeBlock = []string{}
			}
			continue
		}

		if inCode {
			codeBlock = append(codeBlock, line)
			continue
		}

		if line == "---" || line == "----" {
			content = append(content, map[string]interface{}{"type": "rule"})
			continue
		}

		if len(line) > 0 && line[0] == '#' {
			level := 0
			for level < len(line) && line[level] == '#' {
				level++
			}
			if level >= 1 && level <= 6 && level < len(line) && line[level] == ' ' {
				content = append(content, heading(level, line[level+1:]))
				continue
			}
		}

		if strings.HasPrefix(line, "| ") && strings.HasSuffix(line, "|") {
			tableLines := []string{line}
			for i+1 < len(lines) && strings.HasPrefix(lines[i+1], "|") {
				i++
				tableLines = append(tableLines, lines[i])
			}
			table := parseTable(tableLines)
			if table != nil {
				content = append(content, table)
			}
			continue
		}

		if strings.HasPrefix(line, "- ") || strings.HasPrefix(line, "* ") || strings.HasPrefix(line, "1. ") {
			listLines := []string{line}
			for i+1 < len(lines) && (strings.HasPrefix(lines[i+1], "- ") || strings.HasPrefix(lines[i+1], "* ") || strings.HasPrefix(lines[i+1], "1. ") || (len(lines[i+1]) > 2 && lines[i+1][0] >= '0' && lines[i+1][0] <= '9' && strings.Contains(lines[i+1][:3], "."))) {
				i++
				listLines = append(listLines, lines[i])
			}
			ordered := strings.HasPrefix(listLines[0], "1.") || (len(listLines[0]) > 2 && listLines[0][0] >= '0' && listLines[0][0] <= '9')
			list := parseList(listLines, ordered)
			if list != nil {
				content = append(content, list)
			}
			continue
		}

		if strings.TrimSpace(line) == "" {
			continue
		}

		flushParagraph(line)
	}

	if inCode && len(codeBlock) > 0 {
		content = append(content, map[string]interface{}{
			"type": "codeBlock",
			"attrs": map[string]interface{}{
				"language": codeLang,
			},
			"content": []interface{}{
				map[string]interface{}{
					"type": "text",
					"text": strings.Join(codeBlock, "\n"),
				},
			},
		})
	}

	return map[string]interface{}{
		"type":    "doc",
		"version": 1,
		"content": content,
	}
}

func heading(level int, text string) map[string]interface{} {
	return map[string]interface{}{
		"type": "heading",
		"attrs": map[string]interface{}{
			"level": level,
		},
		"content": parseInlineMarks(text),
	}
}

func parseInlineMarks(text string) []interface{} {
	result := []interface{}{}
	if text == "" {
		return result
	}

	i := 0
	current := ""
	for i < len(text) {
		if i < len(text)-1 && text[i] == '*' && text[i+1] == '*' {
			if current != "" {
				result = append(result, textNode(current, nil))
				current = ""
			}
			end := strings.Index(text[i+2:], "**")
			if end >= 0 {
				result = append(result, textNode(text[i+2:i+2+end], []map[string]interface{}{{"type": "strong"}}))
				i = i + 2 + end + 2
				continue
			}
		}
		if text[i] == '`' {
			if current != "" {
				result = append(result, textNode(current, nil))
				current = ""
			}
			end := strings.Index(text[i+1:], "`")
			if end >= 0 {
				result = append(result, textNode(text[i+1:i+1+end], []map[string]interface{}{{"type": "code"}}))
				i = i + 1 + end + 1
				continue
			}
		}
		if text[i] == '[' {
			linkEnd := strings.Index(text[i:], "](")
			if linkEnd >= 0 {
				urlEnd := strings.Index(text[i+linkEnd+2:], ")")
				if urlEnd >= 0 {
					if current != "" {
						result = append(result, textNode(current, nil))
						current = ""
					}
					linkText := text[i+1 : i+linkEnd]
					url := text[i+linkEnd+2 : i+linkEnd+2+urlEnd]
					result = append(result, map[string]interface{}{
						"type": "text",
						"text": linkText,
						"marks": []interface{}{
							map[string]interface{}{
								"type":  "link",
								"attrs": map[string]interface{}{"href": url},
							},
						},
					})
					i = i + linkEnd + 2 + urlEnd + 1
					continue
				}
			}
		}
		_, size := utf8.DecodeRuneInString(text[i:])
		current += text[i : i+size]
		i += size
	}

	if current != "" {
		result = append(result, textNode(current, nil))
	}
	return result
}

func textNode(text string, marks []map[string]interface{}) map[string]interface{} {
	node := map[string]interface{}{
		"type": "text",
		"text": text,
	}
	if len(marks) > 0 {
		m := []interface{}{}
		for _, mark := range marks {
			m = append(m, mark)
		}
		node["marks"] = m
	}
	return node
}

func parseTable(lines []string) map[string]interface{} {
	if len(lines) < 2 {
		return nil
	}
	rows := []interface{}{}
	for i, line := range lines {
		line = strings.TrimPrefix(line, "|")
		line = strings.TrimSuffix(line, "|")
		cells := strings.Split(line, "|")

		if i == 1 {
			allDashes := true
			for _, c := range cells {
				trimmed := strings.TrimSpace(c)
				if trimmed != "" && !strings.HasPrefix(trimmed, "---") && !strings.HasPrefix(trimmed, ":--") && !strings.HasPrefix(trimmed, "--:") {
					allDashes = false
				}
			}
			if allDashes {
				continue
			}
		}

		cellType := "tableCell"
		if i == 0 {
			cellType = "tableHeader"
		}
		rowCells := []interface{}{}
		for _, c := range cells {
			rowCells = append(rowCells, map[string]interface{}{
				"type": cellType,
				"content": []interface{}{
					map[string]interface{}{
						"type":    "paragraph",
						"content": parseInlineMarks(strings.TrimSpace(c)),
					},
				},
			})
		}
		rows = append(rows, map[string]interface{}{
			"type":    "tableRow",
			"content": rowCells,
		})
	}
	return map[string]interface{}{
		"type":    "table",
		"content": rows,
	}
}

func parseList(lines []string, ordered bool) map[string]interface{} {
	listType := "bulletList"
	if ordered {
		listType = "orderedList"
	}
	items := []interface{}{}
	for _, line := range lines {
		text := line
		if strings.HasPrefix(text, "- ") {
			text = strings.TrimPrefix(text, "- ")
		} else if strings.HasPrefix(text, "* ") {
			text = strings.TrimPrefix(text, "* ")
		} else {
			dotIdx := strings.Index(text, ". ")
			if dotIdx > 0 && dotIdx < 4 {
				text = text[dotIdx+2:]
			}
		}
		items = append(items, map[string]interface{}{
			"type": "listItem",
			"content": []interface{}{
				map[string]interface{}{
					"type":    "paragraph",
					"content": parseInlineMarks(text),
				},
			},
		})
	}
	return map[string]interface{}{
		"type":    listType,
		"content": items,
	}
}

// MarkdownToADF converts markdown text to Jira ADF document structure
func MarkdownToADF(markdown string) map[string]interface{} {
	return markdownToADF(markdown)
}
